### PR TITLE
backend-bitbucket: add 'implicit' oauth

### PR DESCRIPTION
**Pull request by cdevienne**
The implementation is a copy/paste of the gitlab implicit oauth code.
Successfully tested with chromium.

The goal is to be able use bitbucket for authentication without a auth server.

### CODE REVIEW

File: packages/netlify-cms-backend-bitbucket/src/AuthenticationPage.js [hunk 1]
```diff
@@ -17,23 +17,49 @@ export default class BitbucketAuthenticationPage extends React.Component {
...
+        app_id: this.props.config.getIn(['backend', 'app_id']),
+        clearHash: this.props.clearHash,
+      });
+      // Complete implicit authentication if we were redirected back to from the provider.
+      this.auth.completeAuth((err, data) => {
+        if (err) {
+          this.setState({ loginError: err.toString() });
+          return;
+        }
+        console.log('Got data:', data);
```
Comment (erquhart) [inline packages/netlify-cms-backend-bitbucket/src/AuthenticationPage.js]: Extra console log.

Comment (cdevienne) [reply packages/netlify-cms-backend-bitbucket/src/AuthenticationPage.js]: Removed it

**Comment by cdevienne** [comment]
Documentation added (I wanted to do it in the first place but did not find the sources... should have look for it harder).

**Comment by Benaiah** [comment]
@erquhart we should note in the documentation that this will only let you log into the CMS for one hour before having to log in again - see the “Refresh Tokens” heading of https://developer.atlassian.com/cloud/bitbucket/oauth-2/

**Comment by erquhart** [comment]
Totally forgot about that issue.

@cdevienne they're saying the implicit grant response includes a refresh token - you can open a PR that utilizes that attempts to use that token when implicit grant users start getting 401 responses?

**Comment by cdevienne** [comment]
Seems to be a less trivial patch, but I'll give it a try.
Thanks for noticing before we ran into it head on!

**Comment by Benaiah** [comment]
@erquhart @cdevienne as the paragraph I copied notes the refresh token is unavailable for the implicit grant auth flow - this issue is unfixable as far as I’m aware with the current state of the BitBucket API, and the issue should be noted in the documentation.

**Comment by erquhart** [comment]
Argh I misread the word "excluded", good catch. Yeah we should just document it.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by cdevienne** [comment]
@Benaiah The implicit grant needs not a refresh token: simply opening the auth page in a hidden iframe will work most of the time, as bitbucket will successfully identify the user without any confirmation needed.

**Comment by erquhart** [comment]
Yeah I was wondering if we could do some kind of workaround behind the scenes, that's a good plan @cdevienne. Let us know if you're planning on implementing that, would love to remove that note from the docs!

**Comment by danielserranoh** [comment]
Hello. I am having and issue trying to understand the documentation to add bitbucket as the backend. According to this doc, the second point reads:

>  Bitbucket gives you a Key. Copy this Key and enter it in your Netlify CMD 

I do not understand two things:

1. Where do I hace to copy the KEY given by Bitbucket
2. What do you mean with "Netlify CMD"

Can you clarify, please?

**Comment by erquhart** [comment]
For help with your implementation, please reach out in our [community chat](https://gitter.im/netlify/netlifycms).

**Comment by mushkin-v** [comment]
@erquhart @cdevienne @Benaiah Still receiving 401 after one hour. Do you have any solution for this issue?

**Pull request merged**